File: baekjoon/16234_populationMoving.py
import sys
import queue

MOVES = [(-1, 0), (0, 1), (1, 0), (0, -1)]

N, L, R = map(int, sys.stdin.readline().rstrip().split())

pop_map = []
for i in range(N):
    pop_row = list(map(int, sys.stdin.readline().rstrip().split()))
    pop_map.append(pop_row)

def is_valid_location(row, col, N):
    if( 0 <= row < N and 0 <= col < N):
        return True
    else:
        return False

def BFS(pop_map, visited, start_row, start_col, group_number):
    next_queue = queue.Queue()
    next_queue.put((start_row, start_col))
    pop_sum = 0
    count = 0
    while(not next_queue.empty()):
        row, col = next_queue.get()
        pop_sum += pop_map[row][col]
        count += 1
        visited[row][col] = group_number
        
        for MOVE in MOVES:
            next_row = row + MOVE[0]
            next_col = col + MOVE[1]
            if(is_valid_location(next_row, next_col, len(pop_map)) and ( L <= abs(pop_map[row][col] - pop_map[next_row][next_col]) <= R) ):
                if(not visited[next_row][next_col]):
                    next_queue.put((next_row, next_col))
                    visited[next_row][next_col] = group_number
    return (pop_sum, count)

move_count = 0
while(True):
    visited = [ [ False for _ in range(N) ] for _ in range(N) ]
    group_number = 1
    groups = [None]
    have_group = False
    for row in range(N):
        for col in range(N):
            if(not visited[row][col]):
                pop_sum, city_count = BFS(pop_map, visited, row, col, group_number)
                if(city_count > 1):
                    have_group = True
                groups.append(int(pop_sum / city_count))
                group_number += 1
    if(have_group):
        move_count += 1
    else:
        break
    for row in range(N):
        for col in range(N):
            group_number = visited[row][col]
            if(group_number):
                pop_map[row][col] = groups[group_number]

print(move_count)

File: baekjoon/16234_populationMoving.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int read_input();
int print_output(long long result);
long long solution();
int N, L, R;
vector<vector<int>> pop_map;

struct Point
{
    int row, col;
};

Point MOVES[] = { {-1, 0}, {0, 1}, {1, 0}, {0, -1} };

int main()
{
    read_input();
    long long result = solution();
    print_output(result);
    return 0;
}

bool is_valid_location(Point point)
{
    if(point.row >= 0 && point.col >= 0 && point.row < N && point.col < N)
        return true;
    else
        return false;
}

pair<int,int> BFS(vector<vector<int>> &visited, Point start_point, int group_number)
{
    queue<Point> next_queue;
    next_queue.push(start_point);
    int pop_sum = 0;
    int count = 0;
    while (!next_queue.empty())
    {
        Point current = next_queue.front();
        next_queue.pop();
        pop_sum += pop_map[current.row][current.col];
        count += 1;
        visited[current.row][current.col] = group_number;

        for(int i = 0; i < 4; i++)
        {
            Point next = {current.row + MOVES[i].row, current.col + MOVES[i].col};
            if(is_valid_location(next))
            {
                int gap = abs(pop_map[current.row][current.col] - pop_map[next.row][next.col]);
                if(gap >= L && gap <= R && !visited[next.row][next.col])
                {
                    next_queue.push(next);
                    visited[next.row][next.col] = group_number;
                }
            }
        }
    }
    return make_pair(pop_sum, count);
}

long long solution()
{
    int move_count = 0;
    while (true)
    {
        vector<vector<int>> visited(N, vector<int>(N, 0));
        int group_number = 1;
        vector<int> avg_groups(1);
        bool have_group = false;

        for (int row = 0; row < N; row++)
        {
            pop_map[row].resize(N);
            for (int col = 0; col < N; col++)
            {
                if (!visited[row][col])
                {
                    pair<int, int> result = BFS(visited, {row, col}, group_number);
                    if(result.second > 1)
                        have_group = true;
                    avg_groups.push_back(result.first / result.second);
                    group_number++;
                }
            }
        }
        if(have_group)
            move_count++;
        else
            break;
        for (int row = 0; row < N; row++)
        {
            for (int col = 0; col < N; col++)
            {
                int group = visited[row][col];
                if(group)
                    pop_map[row][col] = avg_groups[group];
            }
        }
    }
    return move_count;
}

int read_input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cin >> N >> L >> R;
    pop_map.resize(N);
    for (int row = 0; row < N; row++)
    {
        pop_map[row].resize(N);
        for (int col = 0; col < N; col++)
        {
            cin >> pop_map[row][col];
        }
    }
    return 0;
}

int print_output(long long result)
{
    cout << result << "\n";
    return 0;
}